services/tns_submission_queue/tns_submission_queue.py

```diff
@@ -52,6 +52,618 @@ class TNSReportError(Exception):
+        The submission request.
+    session : `~sqlalchemy.orm.Session`
+        The database session to use.
+
+    Returns
+    -------
+    tnsrobot_group : `~skyportal.models.TNSRobotGroup`
+        The TNSRobotGroup to use for submission.
+    tnsrobot_groups : list of `~skyportal.models.TNSRobotGroup`
+        The TNSRobotGroups that the user is allowed to submit to with this robot.
+    """
+    tnsrobot_id = tnsrobot.id
+    user_id = user.id
+
+    tnsrobot_groups = tnsrobot.groups
+    user_accessible_group_ids = [group.id for group in user.accessible_groups]
+    tnsrobot_groups = [
+        tnsrobot_group
+        for tnsrobot_group in tnsrobot_groups
+        if tnsrobot_group.group_id in user_accessible_group_ids
+    ]
+
+    if len(tnsrobot_groups) == 0:
+        raise TNSReportError(
+            f'User {user_id} does not have access to any group with TNSRobot {tnsrobot_id}.'
+        )
+
+    # if this is an auto_submission, we need to check if the group has auto-report enabled
+    # and if the user is an auto-reporter for that group
+    if submission_request.auto_submission:
+        tnsrobot_groups_with_autoreport = [
+            tnsrobot_group
+            for tnsrobot_group in tnsrobot_groups
+            if tnsrobot_group.auto_report is True
+        ]
+        if len(tnsrobot_groups_with_autoreport) == 0:
+            raise TNSReportError(
+                f'No group with TNSRobot {tnsrobot_id} set to auto-report.'
+            )
+
+        # we filter out the groups that this user is not an auto-reporter for
+        tnsrobot_groups_with_autoreport = [
+            tnsrobot_group
+            for tnsrobot_group in tnsrobot_groups_with_autoreport
+            if session.scalar(
+                sa.select(TNSRobotGroupAutoreporter).where(
+                    TNSRobotGroupAutoreporter.tnsrobot_group_id == tnsrobot_group.id,
+                    TNSRobotGroupAutoreporter.group_user_id.in_(
+                        sa.select(GroupUser.id).where(
+                            GroupUser.user_id == user_id,
+                            GroupUser.group_id == tnsrobot_group.group_id,
+                        )
+                    ),
+                )
+            )
+        ]
+
+        if len(tnsrobot_groups_with_autoreport) == 0:
+            raise TNSReportError(
+                f'User {user_id} is not an auto-reporter for any group with TNSRobot {tnsrobot_id} set to auto-report.'
+            )
+
+        # we use the first remaining auto-report group to submit the object
+        tnsrobot_group = tnsrobot_groups_with_autoreport[0]
+    else:
+        # we use the first remaining group to submit the object
+        tnsrobot_group = tnsrobot_groups[0]
+
+    return tnsrobot_group, tnsrobot_groups
+
+
+def apply_existing_tnsreport_rules(tns_headers, tnsrobot, submission_request, session):
+    """Apply the rules for existing TNS reports to the submission request.
```

Comment:
What are these rules defined? Or is this the logic in the function to prevent duplicate submissions?

Reply:
I could describe them in greater detail. here too, but it's described at the DB model schema, as this relies on the `report_existing` flag.